lib/Model/Crawler.php

```diff
@@ -43,14 +45,15 @@ public function crawl(Url $documentUrl, UrlQueue $queue, ReportBuilder $report):
         $xpath = new DOMXPath($dom);
 
         foreach ($xpath->query('//a') as $linkElement) {
+            assert($linkElement instanceof DOMElement);
             $href = $linkElement->getAttribute('href');
 
             if (!$href) {
                 continue;
             }
 
             try {
-                $url = $documentUrl->resolveUrl($href);
+                $url = $documentUrl->resolveUrl($href, ReferringElement::fromDOMElement($linkElement));
```

Comment:
we don't need to pass the link separately here, as it's part of the "referring element". This class needs refactoring.